### DISCUSSION TITLE
IPSEC tunnel not reestablishing to remote SAs after connection reenabled

### DISCUSSION THREAD

**Discussion by davehouser1** [strongswan/strongswan#797] [Q&A]
**Problem:**
IPSEC tunnel not reestablishing to remote SAs after connection re-enabled

**System**
- Centos 8
- Strongswan 5.9.4
- All systems are virtual and running in GNS3
- Remote SA's are Juniper VSRx's

**Overview**
- Strongswan is set up to connect to two remote SAs. If I start the Strongswan  service they both connect and are operational. If I deactivate both or either connection, and wait a period of time (lets say 5 min) then enable the remote SA again, Strongswan will not reestablish the connection. Connections will stay down. Unless I manually tell Strongswan to reconnect. I want strongswan to recognize if the connection is active on the far end and try to reconnect on its own.
- I assumed DPD would send packets constantly even when the connection is down to check when the connection is backup, but I dont think that is the case. I think dpd_delay is only used when the peer goes down. After the peer is removed are any more dpd packets sent? The [swanctl.conf readme](https://wiki.strongswan.org/projects/strongswan/wiki/Swanctlconf) reads like this for dpd_delay:

> Interval to check the liveness of a peer actively using IKEv2 INFORMATIONAL exchanges or IKEv1 R_U_THERE messages. Active DPD checking is only enforced if no IKE or ESP/AH packet has been received for the configured DPD delay.

- Is there a way for Strongswan to recognize when a SA is active again, and reestablish the connection? If not is the only workaround to script some service that is sending its own health packets that will recognize when the peer is active again and then run a `swanctl -i`?
- Note we must use IKEv1, and we must use vti interfaces. 
- Read though [this post](https://serverfault.com/questions/970010/how-can-i-make-strongswan-clients-auto-reconnect-when-server-is-down). @tobiasbrunner  you mention using a trap policy to recognize traffic, and try to re-establish the tunnel. I am having difficulty finding documentation on how to create a trap policy, how would I accomplish this? 

**Configurations**
Below are the configuration files I am using

**swanctl.conf**

```
connections {
    remotesa01 {
        version = 1 
        local_addrs = 2.2.0.2
        remote_addrs = 3.3.0.2
        proposals = aes256-sha256-ecp384
        dpd_delay=1s
        dpd_timeout=5s
        local {
            id = 2.2.0.2
            auth = psk 
        }
        remote {
            auth = psk 
        }
        children {
            remotesa01-child {
                local_ts = 0.0.0.0/0
                remote_ts = 0.0.0.0/0
                mark_in = 41
                mark_out = 41
                esp_proposals = aes256-sha256-ecp384
                updown = /opt/_updown_vti01 iptables
                start_action = start
                close_action = start
                dpd_action = restart
            }
        }
    }   
    remotesa02 { 
        version = 1 
        local_addrs = 2.2.0.2 
        remote_addrs = 3.3.1.2 
        proposals = aes256-sha256-ecp384 
        dpd_delay=1s 
        dpd_timeout=5s 
        local { 
            id = 2.2.0.2 
            auth = psk 
        } 
        remote { 
            auth = psk 
        } 
        children {
            remotesa02-child {
                local_ts = 0.0.0.0/0
                remote_ts = 0.0.0.0/0
                mark_in = 42
                mark_out = 42
                esp_proposals = aes256-sha256-ecp384
                updown = /opt/_updown_vti02 iptables
                start_action = start
                close_action = start
                dpd_action = restart
            }
        }
    }
}
secrets {
    ike-me {
        secret = "password123"
        id-conn01 = 2.2.0.2
    }
    ike-remotesa01 {
        secret = "password123"
        id-to-remotesa01 = 3.3.0.2
    }
    ike-remotesa02 {
        secret = "password123"
        id-to-remotesa02 = 3.3.1.2
    }
}
include conf.d/*.conf
```

**_updown_vti01**

```
#!/bin/bash

set -o nounset
set -o errexit

VTI_IF="vti01"
VTI_LOCAL_ADDRESS="172.21.0.2"
CUSTOM_MY_CLIENT="172.21.0.0/29"

case "${PLUTO_VERB}" in
    up-client)
        ip tunnel add $VTI_IF local ${PLUTO_ME} remote ${PLUTO_PEER} mode vti key 41
        ip link set $VTI_IF up
        ip addr add  $VTI_LOCAL_ADDRESS dev $VTI_IF
        ip route add ${CUSTOM_MY_CLIENT} dev $VTI_IF
        sysctl -w "net.ipv4.conf.$VTI_IF.disable_policy=1"
        sysctl -w "net.ipv4.conf.$VTI_IF.rp_filter=0"
        sysctl -w "net.ipv4.conf.$VTI_IF.forwarding=1"
        ;;
    down-client)
        ip tunnel del $VTI_IF
        ;;
esac
```

**_updown_vit02**

```
#!/bin/bash

set -o nounset
set -o errexit

VTI_IF="vti02"
VTI_LOCAL_ADDRESS="172.22.0.2"
CUSTOM_MY_CLIENT="172.22.0.0/29"
CUSTOM_PEER_CLIENT="10.0.26.0/24"

case "${PLUTO_VERB}" in
    up-client)
        ip tunnel add $VTI_IF local ${PLUTO_ME} remote ${PLUTO_PEER} mode vti key 42
        ip link set $VTI_IF up
        ip addr add  $VTI_LOCAL_ADDRESS dev $VTI_IF
        ip route add ${CUSTOM_MY_CLIENT} dev $VTI_IF
        sysctl -w "net.ipv4.conf.$VTI_IF.disable_policy=1"
        sysctl -w "net.ipv4.conf.$VTI_IF.rp_filter=0"
        sysctl -w "net.ipv4.conf.$VTI_IF.forwarding=1"
        ;;
    down-client)
        ip tunnel del $VTI_IF
        ;;
esac
```

**Comment by tobiasbrunner** [comment]
Sorry, I don't really understand your description. There is a confusion between "connection", "SA" and possibly "config" if you refer to entries in the "connections" section with "connection" in some instances.

> Note we must use IKEv1, and we must use vti interfaces.

Nobody _must_ use IKEv1 nowadays, they are either lazy or cheap (or stupid). And using VTIs is basically a local decision, even a route-based approach does not necessarily require additional interfaces. And on newer kernels, XFRM interfaces are a way better choice (they are available since 4.19, and while the CentOS 8 kernel is officially 4.18, support for them might have been backported).

> * I am having difficulty finding documentation on how to create a trap policy, how would I accomplish this?

How so? Where did you look? But you can't do that anyway if you use VTIs that are created in an updown script (not sure if it would work if you created them and the routes before, however, that does work with XFRM interfaces).

**Reply by davehouser1** [reply]
@tobiasbrunner 

> Sorry, I don't really understand your description. There is a confusion between "connection", "SA" and possibly "config" if you refer to entries in the "connections" section with "connection" in some instances.

I updated the "Overview" section of my post, I also referenced an older post to show the topology. Sorry if my description is confusing. Let me know if you need more clarification. 

Re: IKEv1, that was a business decision by the company, I have no control of it, its just a requirement that they use. 

Re: XFRM back port, how can I tell if XFRM is backported to 4.18? I did some light searching online but was not able to definitively tell if it was or not. Is there any command to check if xfrm actually works in my version of 4.18?

Re: trap documentation, What are the steps to creating a trap? I see in the [swanctl.conf file documentation ](https://wiki.strongswan.org/projects/strongswan/wiki/Swanctlconf) you can apply "trap" to start_action or close_action, however I am not sure I understand if there are other configurations need to be done to trap specific traffic, and open the connection if that traffic is recognized. I thought a trap would recognize traffic based on protocol, or subnet, or other factors, then open up the interface. Is my assumption wrong? is there any detailed documentation I can read though on the step by step process to create a trap with xfrm interfaces?

**Reply by tobiasbrunner** [reply]
> I updated the "Overview" section of my post, I also referenced an older post to show the topology.

Hm, I don't see any changes (says last edited 3 days ago).

> Re: IKEv1, that was a business decision by the company, I have no control of it, its just a requirement that they use.

Then quit, that's a horrible decision.

> Is there any command to check if xfrm actually works in my version of 4.18?

See [RouteBasedVPN](https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN#XFRM-Interfaces-on-Linux) (either the `ip link` command, or if that's too old and doesn't know `type xfrm`, via strongSwan's `xfrmi` command).

> I thought a trap would recognize traffic based on protocol, or subnet, or other factors, then open up the interface. Is my assumption wrong?

No, that's exactly it. If you configure `start_action = trap`, IPsec policies will be installed in the kernel, based on `local|remote_ts` and the marks and interface IDs. Then if traffic matches these, an acquire will be generated by the kernel that triggers the creation of IKE and IPsec SAs. But that's obviously trickier with route-based VPNs that involve an interface. Traffic won't match the policies without those interfaces (while you could mark traffic manually if you use VTIs, kinda defeating the whole purpose of the interface, with XFRM interfaces that's not possible as the policies will only match packets going through a particular interface). Creating interfaces beforehand is easy for XFRM interfaces as you don't need any addresses, with VTIs, at least one address has to be set and only one interface can omit the remote address for the same local address (see the above page for details).
